flax/linen/linear.py

```diff
@@ -199,56 +201,22 @@ def _conv_dimension_numbers(input_shape):
   return lax.ConvDimensionNumbers(lhs_spec, rhs_spec, out_spec)
 
 
-class Conv(Module):
-  """Convolution Module wrapping lax.conv_general_dilated.
+def _get_conv_call(shared_weights: bool) -> Callable:
```

Comment:
why not have a shared_weights arg on Conv and perhaps a `LocalConv = partial(Conv, shared_weights=False)`?

Reply:
I agree it would be cleaner, but when I try
```python
class _Conv(Module):
  """...
  """
  shared_weights: bool
  features: int
  kernel_size: Iterable[int]
  strides: Union[None, int, Iterable[int]] = 1
  padding: Union[str, Iterable[Tuple[int, int]]] = 'SAME'
  input_dilation: Union[None, int, Iterable[int]] = 1
  kernel_dilation: Union[None, int, Iterable[int]] = 1
  feature_group_count: int = 1
  use_bias: bool = True
  dtype: Dtype = jnp.float32
  param_dtype: Dtype = jnp.float32
  precision: Any = None
  kernel_init: Callable[[PRNGKey, Shape, Dtype], Array] = default_kernel_init
  bias_init: Callable[[PRNGKey, Shape, Dtype], Array] = zeros

  ...


Conv = functools.partial(_Conv, True)


ConvLocal = functools.partial(_Conv, False)
```
I get `flax:checkpoints_test` failure as
```python
Traceback (most recent call last):
  File "/build/work/710b6a4d0078c1beccde29236a65e8895303/google3/runfiles/google3/third_party/py/flax/tests/checkpoints_test.py", [line 312](https://cs.corp.google.com/piper///depot/google3/third_party/py/flax/tests/checkpoints_test.py?l=312&ws=romann/2404&snapshot=239), in test_convert_checkpoint
    checkpoints.convert_pre_linen(state_dict['optimizer'])))
  File "/build/work/710b6a4d0078c1beccde29236a65e8895303/google3/runfiles/google3/third_party/py/flax/optim/base.py", [line 350](https://cs.corp.google.com/piper///depot/google3/third_party/py/flax/optim/base.py?l=350&ws=romann/2404&snapshot=239), in restore_state
    self.target, self.state, state)
  File "/build/work/710b6a4d0078c1beccde29236a65e8895303/google3/runfiles/google3/third_party/py/flax/optim/base.py", [line 155](https://cs.corp.google.com/piper///depot/google3/third_party/py/flax/optim/base.py?l=155&ws=romann/2404&snapshot=239), in restore_state
    opt_target = serialization.from_state_dict(opt_target, state_dict['target'])
  File "/build/work/710b6a4d0078c1beccde29236a65e8895303/google3/runfiles/google3/third_party/py/flax/serialization.py", [line 51](https://cs.corp.google.com/piper///depot/google3/third_party/py/flax/serialization.py?l=51&ws=romann/2404&snapshot=239), in from_state_dict
    return ty_from_state_dict(target, state)
  File "/build/work/710b6a4d0078c1beccde29236a65e8895303/google3/runfiles/google3/third_party/py/flax/core/frozen_dict.py", [line 199](https://cs.corp.google.com/piper///depot/google3/third_party/py/flax/core/frozen_dict.py?l=199&ws=romann/2404&snapshot=239), in _restore_frozen_dict
    for key, value in xs.items()})
  File "/build/work/710b6a4d0078c1beccde29236a65e8895303/google3/runfiles/google3/third_party/py/flax/core/frozen_dict.py", [line 199](https://cs.corp.google.com/piper///depot/google3/third_party/py/flax/core/frozen_dict.py?l=199&ws=romann/2404&snapshot=239), in <dictcomp>
    for key, value in xs.items()})
KeyError: '_Conv_0'
```
So it looks like there's some naming issue where the "children" names `Conv` and `ConvLocal` aren't used, and `_Conv` is used, do you have ideas how to fix it?

Reply:
It might be easier to use subclassing here. You could make shared_weights an abstract property of the base class. `partial` indeed gives issues with class names (and documentation rendering)

Reply:
Thanks for the suggestion, done--ptal!